File: back/src/main/java/org/bibliotheque/bibliotheque/modele/entity/Livre.java
package org.bibliotheque.bibliotheque.modele.entity;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;
import org.bibliotheque.bibliotheque.util.enumm.Etat;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.Arrays;
import java.util.List;

@AllArgsConstructor
@NoArgsConstructor
@Data
@SuperBuilder
@Entity
@Table
public class Livre {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id_livre")
    private Integer idLivre;

    @NotNull
    @NotBlank(message = "Le titre ne doit pas être vide")
    private String titre;

    @NotNull
    @NotBlank(message = "Le description ne doit pas être vide")
    private String description;

    @Enumerated(EnumType.STRING)
    @NotNull
    private Etat etat;

    @NotNull
    private int nbrPage;  // Since it's an int, no need for @NotBlank.

    @Min(value = 0, message = "Le nombre d'étoiles ne peut pas être inférieur à 0")
    @Max(value = 5, message = "Le nombre d'étoiles ne peut pas être supérieur à 5")
    private int star;  // The rating must be between 0 and 5.

    @Lob
    @Column(name = "image", columnDefinition = "LONGBLOB")
    @NotNull
    private byte[] image;  // Storing image as a byte array.

    private Boolean statut;


    @CreationTimestamp
    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    @UpdateTimestamp
    private LocalDateTime updatedAt;

    @Column(name = "deleted_at")
    private LocalDateTime deletedAt;

    @Column(name = "auther_id")
    private Integer idAuther;

    // Relationship with Auther
    @JsonBackReference("livre_auther")
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "auther_id", insertable = false, updatable = false)
    private Auther auther;

    // Relationship with Reservation
    @JsonManagedReference("reservation_livre")
    @OneToMany(mappedBy = "livre", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Reservation> reservations;

    // Relationship with Emprunt
    @JsonManagedReference("emprunt_livre")
    @OneToMany(mappedBy = "livre", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Emprunt> emprunts;

    @Override
    public String toString() {
        return "Livre{" +
                "idLivre=" + idLivre +
                ", titre='" + titre + '\'' +
                ", description='" + description + '\'' +
                ", etat=" + etat +
                ", nbrPage=" + nbrPage +
                ", star=" + star +
                ", image=" + Arrays.toString(image) +
                ", statut=" + statut +
                ", createdAt=" + createdAt +
                ", updatedAt=" + updatedAt +
                ", deletedAt=" + deletedAt +
                ", idAuther=" + idAuther +
                '}';
    }
}

File: web/src/app/shared/models/livre.ts
import { stat } from "fs";

export class Livre {
  private _idLivre?: number;
  private _titre: string;
  private _nbrPage: number;
  private _description: string;
  private _star: number;
  private _image: string;
  private  _etat: string | undefined;

  private _statut:boolean;

  private _idAuther?: number;
  private _hidden?: boolean; // Make it optional, as it's not a required field from the database.

  constructor(
    titre: string,
    nbrPage: number,
    description: string,
    image: string,

    star: number,
    idAuther?: number,
    idLivre?: number,
    statut: boolean=false,

  ) {
    this._titre = titre;
    this._nbrPage = nbrPage;
    this._description = description;
    this._image = image;

    this._star = star;
    this._idAuther = idAuther;
    this._idLivre = idLivre;
    this._statut=statut;
  }



  // Getters and Setters
  get idLivre(): number | undefined {
    return this._idLivre;
  }

  set idLivre(idLivre: number) {
    this._idLivre = idLivre;
  }

  get titre(): string {
    return this._titre;
  }

  set titre(titre: string) {
    this._titre = titre;
  }

  get nbrPage(): number {
    return this._nbrPage;
  }

  set nbrPage(nbrPage: number) {
    this._nbrPage = nbrPage;
  }

  get description(): string {
    return this._description;
  }

  set description(description: string) {
    this._description = description;
  }

  get etat(): string | undefined{
    return this._etat;
  }

  set etat(description: string) {
    this._etat = description;
  }

  get image(): string {
    return this._image;
  }

  set image(image: string) {
    this._image = image;
  }

  get star(): number {
    return this._star;
  }

  set star(star: number) {
    this._star = star;
  }

  get idAuther(): number | undefined {
    return this._idAuther;
  }

  set idAuther(idAuther: number) {
    this._idAuther = idAuther;
  }

  // Correct getter and setter for hidden
  get hidden(): boolean {
    return this._hidden ?? false; // Return false if _hidden is undefined
  }

  set hidden(hidden: boolean) {
    this._hidden = hidden;
  }

  get statut(): boolean {
    return this._statut ?? false; // Return false if _hidden is undefined
  }

  set statut(hidden: boolean) {
    this._statut = hidden;
  }
}
